doc/sphinx/proof-engine/tactics.rst

```diff
@@ -102,35 +101,40 @@ specified, the default selector is used.
 Bindings
 ~~~~~~~~
 
-Tactics that take a term as an argument may also accept :token:`bindings`
-to instantiate some parameters of the term by name or position.
-The general form of a term with :token:`bindings` is
-:n:`@term__tac with @bindings` where :token:`bindings` can take two different forms:
+Tactics that take a term as an argument may also accept :token:`bindings` to
+specify the values to assign unbound variables in a term.
+Bindings can be given by position or name.  Generally these appear in the form
+:n:`with @bindings`.
```

Comment:
Since the text says "these appear in the form...", I think it would make sense to have the `one_term_with_bindings` appear just below. But do as you prefer.

Reply:
I left it as-is.  There are a bunch of tactics that have `with bindings` that don't use `one_term_with_bindings`, i.e.

```
| "elim" one_term_with_bindings OPT ( "using" one_term OPT ( "with" bindings ) )
| "specialize" one_term OPT ( "with" bindings ) OPT ( "as" simple_intropattern )
| "left" OPT ( "with" bindings )
| "right" OPT ( "with" bindings )
| "econstructor" OPT ( nat_or_var OPT ( "with" bindings ) )
| "setoid_rewrite" OPT [ "->" | "<-" ] one_term OPT ( "with" bindings ) OPT ( "at" rewrite_occs ) OPT ( "in" ident )
| "soft" "functional" "induction" LIST1 one_term OPT ( "using" one_term OPT ( "with" bindings ) ) OPT ( "as" simple_intropattern )      (* funind plugin *)
| "setoid_rewrite" OPT [ "->" | "<-" ] one_term OPT ( "with" bindings ) "in" ident "at" rewrite_occs
| "substitute" OPT [ "->" | "<-" ] one_term OPT ( "with" bindings )
| "eelim" one_term_with_bindings OPT ( "using" one_term OPT ( "with" bindings ) )
| "eleft" OPT ( "with" bindings )
| "contradiction" OPT ( one_term OPT ( "with" bindings ) )
| "constructor" OPT nat_or_var OPT ( "with" bindings )
| "functional" "induction" term OPT ( "using" one_term OPT ( "with" bindings ) ) OPT ( "as" simple_intropattern )      (* funind plugin *)
| "using" one_term OPT ( "with" bindings ) OPT occurrences
| "eright" OPT ( "with" bindings )
| "split" OPT ( "with" bindings )
| "esplit" OPT ( "with" bindings )
```

which is different from `one_term_with_bindings` 

```
one_term_with_bindings: [
| OPT ">" one_term OPT ( "with" bindings )
]
```

Do you think some of these tactics should be generalized to use `one_term_with_bindings`?  BTW, what does the `">"` mean, I can add that with `one_term_with_bindings`.

Reply:
I removed "Refine passed ill-formed term." and "Invalid argument." exceptions, I don't see them in the code, either.

Reply:
`fullGrammar` has `constr_with_bindings` which is exactly `one_term OPT ( "with" bindings )` and `constr_with_bindings_arg`, which is what has become `one_term_with_bindings` (following [this short discussion](https://github.com/coq/coq/pull/13470#discussion_r531680189) where we didn't consider the two cases), includes the `OPT ">"`. Maybe we should rebind `one_term_with_bindings` to `constr_with_bindings` to be able to use it pervasively...

Reply:
According to this test:

```
Goal (True -> True) -> True.
intro H.
apply > H.
```

and this error message:

```
keep/clear modifiers apply only to hypothesis names
```

`>` is a keep/clear modifier, meaning that it can clear a hypothesis name when it is applied in tactics that would keep it otherwise (`apply`) and keep it in tactics that would remove it otherwise (`destruct`).

So really, it should not be kept part of the `one_term_with_bindings` form (and should probably be spliced instead).

Reply:
BTW, I wonder if this modifier is actually used in practice and if it could make sense to remove it instead of documenting it.

Reply:
Updated.  I temporarily removed the `>` production to see whether CI relies on it.  It's a bit odd to have `>` mean opposite things depending on where it's used.  I'll consider updating how this is presented shortly.

Reply:
It looks like there are no users in the whole CI. I propose to remove this feature in a separate PR, so that it can be discussed separately, with anyone who wants to.

Reply:
I'll leave it to you to create that PR.  (Don't forget uses of '>' in `induction_arg` and `oriented_rewriter`.

Reply:
See #16407. For now the deprecation is exactly where you had commented out code, so this should be fine, right?

Reply:
> BTW, I wonder if this modifier is actually used in practice and if it could make sense to remove it instead of documenting it.

It is not documented but I believe it to be both convenient in practice and logically motivated: it allows to enforce on the fly (that is w/o using `clear`) that an hypothesis is used linearly, and at the same time it lightens the context for the rest of the proof from now useless hypotheses.

I did not document (nor even initiated a discussion about it) because (chat échaudé craint l'eau froide) I feared that some ssreflect people (again) opposed to transfer "cool" ideas from ssreflect to other tactics. I agree that this is a pity... (scientifically speaking, there is no other road than aiming at some convergence).

Reply:
Also,t here might be better concrete syntax for it.

Reply:
What is the equivalent syntax in SSReflect?

Reply:
> (chat échaudé craint l'eau froide) I feared that some ssreflect people (again) opposed to transfer "cool" ideas from ssreflect to other tactics. I agree that this is a pity... (scientifically speaking, there is no other road than aiming at some convergence).

If a tree falls in the forest and no one is around to hear it, does it make a sound?  Undocumented features don't exist :-).

Seeking some convergence between regular tactics and SSR sounds like a worthy endeavor.

Reply:
Thanks for explaining.